.github/workflows/go.yml

```diff
@@ -0,0 +1,28 @@
+# This workflow will build a golang project
+# For more information see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-go
+
+name: Go
+
+on:
+  push:
+    branches: [ "main" ]
+  pull_request:
+    branches: [ "main" ]
+
+jobs:
+
+  build:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v3
+
+    - name: Set up Go
+      uses: actions/setup-go@v4
+      with:
+        go-version: '1.22.1'
+
+    - name: Build
```

Comment:
Excellent!

I believe we should run the tests before building. Do you have a contrary argument?

Reply:
Please do the change:

In a Go pipeline, the sequence of go build or go test can depend on the objectives of the pipeline and the specifics of the project. Both approaches have their merits:

go test before go build:

Immediate Feedback on Code Quality: Running go test before go build ensures that you receive immediate feedback on the correctness of your code. If tests fail, you know there's a problem that needs to be addressed before proceeding further. This can save time and resources, especially in continuous integration (CI) environments where you want to catch issues early.
Encourages Test-Driven Development (TDD): This approach aligns well with test-driven development practices, where writing and passing tests precede writing new code or refactoring existing code.
Prevents Building Faulty Code: It prevents the pipeline from spending time on building potentially faulty or broken code. If the tests fail, there's no point in proceeding to the build stage.
go build before go test:

Quick Compilation Check: Running go build first is a quick way to check that the code compiles without errors. This might be useful in situations where you want to ensure that the codebase is in a buildable state before spending time running tests, especially if the tests are time-consuming.
Dependencies Check: It ensures that all dependencies are correctly configured and can be compiled. This can be particularly useful if your project includes C bindings or other components that require a successful build process to validate their integration.
Artifact Generation: In some workflows, generating the build artifact even if tests fail later could be useful for certain types of analysis or for allowing developers to proceed with manual testing in parallel with fixing test failures.
Best Practice
The common and generally recommended approach is to run go test before go build in the pipeline. This prioritizes the identification and correction of errors early in the development cycle, ensuring code quality and functionality before the creation of a build artifact. However, specific requirements or workflows might necessitate a different order or an additional setup that runs both commands in parallel or in different scenarios. The key is to adapt the pipeline to best suit the project's needs and the team's workflow.